Reordered some paragraphs, clean up + LightInject example (#1170)
umbraco/UmbracoDocs: Reference/using-ioc.md

In most IoC frameworks you would setup your container in your global.asax class. To do that in Umbraco, you will need to inherit from our global.asax class called: `Umbraco.Web.UmbracoApplication`. You should then override the `OnApplicationStarted` method to build your container and initialize any of the IoC stuff that you require.
Alternatively you can implement the `Umbraco.Web.IApplicationEventHandler` interface.

## What assemblies and controllers do I need to register?

> You need to register all assemblies that may contain MVC or WebApi controllers. In Umbraco this is the `umbraco` assembly which you can get a direct assembly reference to using the example syntax used above:

Reviewer: ```suggestion
You need to register all assemblies that may contain MVC or WebApi controllers. In Umbraco this is the `umbraco` assembly which you can get a direct assembly reference to using the example syntax used below:
```
It's below, right? 🙂

Reviewer 2: @jmayntzhusen yes i think so :) // @mzajkowski

```
typeof(UmbracoApplication).Assembly
```

If you don't register assemblies that contain controllers you may end up with YSOD errors. If you do not register a controller then ASP.NET will try to create the controller but if it doesn't have an empty constructor you'll get a YSOD.

## Things to note

> We use a custom MVC controller builder in our code called `Umbraco.Web.Mvc.MasterControllerFactory`, which needs to always be the default controller factory, if you change this Umbraco will probably not work anymore. The good news is that you can specify 'slave' factories so you can specify custom controller factories for different purposes. You would just need to create a new class that inherits from `Umbraco.Web.Mvc.IFilteredControllerFactory` and ensure that the class is public (so it can be found). If your IoC implementation affects the default controller factory, you may have to modify it in order to support this implementation. For the most part, most IoC frameworks will just target setting a custom DependencyResolver which is 100% ok.

Reviewer 2: We'll need to change the term 'slave' to 'sub' (or similar)

Now create the following files to configure Unity correctly for an Umbraco site.

> 	public class UnityEvents : ApplicationEventHandler

Reviewer 2: This is incorrect, there is a reason why it is using the interface! If you use the base class, then you need to override both: `ExecuteWhenApplicationNotConfigured` and `ExecuteWhenDatabaseNotConfigured` and return true for both of these, by default this is false because this base class is generally just used for plugins/extensions but your container must always be populated.

	public class LightInjectEvents : ApplicationEventHandler
	{
        protected override void ApplicationStarted(UmbracoApplicationBase umbracoApplication, ApplicationContext applicationContext)
		{
            var container = new ServiceContainer();

            // Our services registrations here... 
            // Documentation: https://www.lightinject.net/

            // Registration of the Controllers and WebApi Controllers from the executing assemply
            container.RegisterControllers(Assembly.GetExecutingAssembly());
            container.RegisterApiControllers(Assembly.GetExecutingAssembly());

            // Enabling required configuration and settings
            container.EnableMvc();
            container.EnablePerWebRequestScope();
            container.EnableWebApi(GlobalConfiguration.Configuration);

            // Setting up the resolver
            var resolver = new LightInjectWebApiDependencyResolver(container);
            GlobalConfiguration.Configuration.DependencyResolver = resolver;
            DependencyResolver.SetResolver(new LightInjectMvcDependencyResolver(container));
		}
	}
    

> If we're going to call/implement this class outside of the Web application or inside of the package for example, it might be useful (and sometimes required) to also register all the other existing Controllers and API Controllers inside of the whole solution. Especially if we'll be willing to use them or inject them in our own registrations. Below code is scanning through the assemblies within the current domain and registering existing controllers which are implementing IHttpController interface.

Reviewer: ```suggestion
If we're going to call/implement this class outside of the Web application or inside of the package for example, it might be useful (and sometimes required) to also register all the other existing Controllers and API Controllers inside of the whole solution. Especially if we'll be willing to use them or inject them in our own registrations. The code below scans through the assemblies within the current domain and registering existing controllers which are implementing the IHttpController interface.
```

Reviewer 2: @mzajkowski agreed with the above request

        private static void RegisterAssembliesServices(IServiceRegistry container)
        {
            foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies())
            {
                var controllerTypes = assembly.GetTypes().Where(t => !t.IsAbstract && typeof(IHttpController).IsAssignableFrom(t));
                foreach (var controllerType in controllerTypes)
                {
                    container.Register(controllerType, new PerRequestLifeTime());
                }
            }
        }
        
We can call the helper method after our own services registrations.

## Our.Umbraco.IoC

> There is also another way of dealing with IoC containers in Umbraco solution and it's with using the package created by one of the Core developers - Shannon Deminick - called: Our.Umbraco.IoC (https://github.com/Shazwazza/Our.Umbraco.IoC).

Reviewer: ```suggestion
There is also another way of dealing with IoC containers in Umbraco solutions and it's with using the package created by one of the Core developers - Shannon Deminick - called: [Our.Umbraco.IoC](https://github.com/Shazwazza/Our.Umbraco.IoC)
```

> There is also another way of dealing with IoC containers in Umbraco solution and it's with using the package created by one of the Core developers - Shannon Deminick - called: Our.Umbraco.IoC (https://github.com/Shazwazza/Our.Umbraco.IoC).

Reviewer 2: I think this should go at the very top, people don't need to manually implement all of this and it would save a lot of headaches to just use this as the recommended approach.
